File: system/fast_risk_cache.py
"""
system/fast_risk_cache.py
DIX VISION v42.2 — Precomputed Risk Cache (FAST PATH, NO RPC)

Updated asynchronously by Governance.
Consumed synchronously by Indira with zero latency.
Thread-safe reads via atomic reference swap.

# Backends

This module has two interchangeable backends and picks one at import time:

* **Rust (``dixvision_py_system``)** — PyO3 extension built from
  ``rust/py_system/``. Preferred when the wheel is installed. Read
  path is a lock-free ``ArcSwap::load()`` on the Rust side; the
  Python side only pays a single FFI trampoline + tuple marshalling
  per call and reconstructs the ``RiskConstraints`` dataclass locally.
* **Pure Python (fallback)** — the original CPython-atomic-reference
  implementation kept verbatim so pre-polyglot ledgers still replay
  on a box without the Rust wheel (no cargo, restricted glibc, etc).

Both backends satisfy the same invariants:

* ``get()`` returns an immutable ``RiskConstraints`` snapshot without
  ever blocking a concurrent writer (single-writer, multi-reader).
* ``update(**kwargs)`` serialises writes and atomically swaps in a
  fresh snapshot, stamping ``last_updated_utc`` to the current
  UTC-ISO string.
* ``enter_safe_mode`` / ``exit_safe_mode`` / ``halt_trading`` /
  ``resume_trading`` are convenience ``update()`` calls with the
  same semantics on both sides.
* ``allows_trade(size_usd, portfolio_usd)`` is a pure function on the
  ``RiskConstraints`` dataclass and therefore backend-independent.

Selection is observable via ``backend()``. Tests in
``tests/test_fast_risk_cache_parity.py`` assert both backends meet
the invariants by constructing instances via ``make_python_cache()``
and ``make_rust_cache()`` directly, without reloading this module
or mutating environment variables.
"""
from __future__ import annotations

from dataclasses import dataclass, replace
from threading import RLock
from typing import Optional

from system.time_source import utc_now


@dataclass(frozen=True)
class RiskConstraints:
    """Precomputed risk limits. Consumed by Indira fast path."""
    max_position_pct: float = 1.0       # max position as % of portfolio
    max_order_size_usd: float = 10_000.0
    volatility_band_high: float = 0.05  # 5%
    volatility_band_low: float = 0.001
    circuit_breaker_drawdown: float = 0.04  # 4%
    circuit_breaker_loss_pct: float = 0.01  # 1% per trade
    trading_allowed: bool = True
    safe_mode: bool = False
    last_updated_utc: str = ""

    def allows_trade(self, size_usd: float, portfolio_usd: float) -> tuple[bool, str]:
        if not self.trading_allowed:
            return False, "trading_not_allowed"
        if self.safe_mode:
            return False, "safe_mode_active"
        # Fail-closed: if we don't know the portfolio size we cannot
        # enforce the per-trade circuit breaker, so refuse rather than
        # silently skipping the percentage check.
        if portfolio_usd <= 0:
            return False, "portfolio_usd_required"
        # Absolute per-order cap governance sets via ConstraintCompiler.
        # Checked BEFORE the percentage rule so a large absolute size
        # on a very large portfolio still gets rejected.
        if size_usd > self.max_order_size_usd:
            return False, (
                f"size_usd_{size_usd:.2f}_exceeds_max_"
                f"{self.max_order_size_usd:.2f}"
            )
        pct = size_usd / portfolio_usd
        if pct > self.circuit_breaker_loss_pct:
            return False, f"size_pct_{pct:.4f}_exceeds_limit_{self.circuit_breaker_loss_pct}"
        return True, "ok"


# ---------------------------------------------------------------------------
# Pure-Python backend. Also the reference implementation the parity
# suite validates the Rust port against.
# ---------------------------------------------------------------------------


class _PythonFastRiskCache:
    """CPython atomic-reference implementation. Attribute assignment on
    a Python object is effectively atomic under the GIL, so readers
    never see a half-built ``RiskConstraints``."""

    __slots__ = ("_constraints", "_lock")

    def __init__(self) -> None:
        self._constraints = RiskConstraints(
            last_updated_utc=utc_now().isoformat()
        )
        self._lock = RLock()

    def get(self) -> RiskConstraints:
        """Lock-free read (atomic reference in CPython)."""
        return self._constraints

    def update(self, **kwargs: object) -> RiskConstraints:
        """Governance calls this asynchronously to update constraints."""
        with self._lock:
            self._constraints = replace(
                self._constraints,
                last_updated_utc=utc_now().isoformat(),
                **kwargs,  # type: ignore[arg-type]
            )
            return self._constraints

    def enter_safe_mode(self) -> RiskConstraints:
        return self.update(safe_mode=True, trading_allowed=False)

    def exit_safe_mode(self) -> RiskConstraints:
        return self.update(safe_mode=False, trading_allowed=True)

    def halt_trading(self, reason: str = "") -> RiskConstraints:
        del reason  # Reason is logged by callers; cache just records the flag.
        return self.update(trading_allowed=False)

    def resume_trading(self) -> RiskConstraints:
        return self.update(trading_allowed=True, safe_mode=False)


def make_python_cache() -> _PythonFastRiskCache:
    """Construct a fresh Python-backed risk cache with its own state.
    Callers: the module-level default when the Rust wheel is
    unavailable, and the parity test suite."""
    return _PythonFastRiskCache()


# ---------------------------------------------------------------------------
# Rust backend wrapper. Imported lazily so a missing wheel falls
# through cleanly to the Python backend without an ``ImportError``
# at module load time.
# ---------------------------------------------------------------------------


# Field order MUST match ``rust/py_system/src/lib.rs::RiskTuple``.
# Any reshuffle breaks the ABI — treat this list and the Rust type
# alias as one diff.
_RUST_TUPLE_FIELDS = (
    "max_position_pct",
    "max_order_size_usd",
    "volatility_band_high",
    "volatility_band_low",
    "circuit_breaker_drawdown",
    "circuit_breaker_loss_pct",
    "trading_allowed",
    "safe_mode",
    "last_updated_utc",
)


def _from_rust_tuple(t: tuple) -> RiskConstraints:
    """Reconstruct a ``RiskConstraints`` from the 9-tuple returned by
    the PyO3 seam. Keep this helper symmetrical with the Rust
    ``snapshot_to_tuple`` function."""
    return RiskConstraints(**dict(zip(_RUST_TUPLE_FIELDS, t)))


def _to_rust_patch(
    max_position_pct: Optional[float] = None,
    max_order_size_usd: Optional[float] = None,
    volatility_band_high: Optional[float] = None,
    volatility_band_low: Optional[float] = None,
    circuit_breaker_drawdown: Optional[float] = None,
    circuit_breaker_loss_pct: Optional[float] = None,
    trading_allowed: Optional[bool] = None,
    safe_mode: Optional[bool] = None,
) -> tuple:
    """Marshal Python kwargs → positional 8-tuple for ``risk_update``.
    ``None`` slots leave the corresponding field unchanged."""
    return (
        max_position_pct,
        max_order_size_usd,
        volatility_band_high,
        volatility_band_low,
        circuit_breaker_drawdown,
        circuit_breaker_loss_pct,
        trading_allowed,
        safe_mode,
    )


class _RustFastRiskCache:
    """Thin wrapper delegating to the process-global Rust cache.

    The underlying ``FastRiskCache`` on the Rust side is a process
    singleton (``OnceLock`` inside the PyO3 seam). Constructing
    multiple instances of this wrapper therefore does NOT give you
    independent caches; tests that need isolation use the Python
    backend via ``make_python_cache()``."""

    __slots__ = ("_rust",)

    def __init__(self, rust_module: object) -> None:
        self._rust = rust_module

    def get(self) -> RiskConstraints:
        return _from_rust_tuple(self._rust.risk_get())  # type: ignore[attr-defined]

    def update(self, **kwargs: object) -> RiskConstraints:
        patch = _to_rust_patch(**kwargs)  # type: ignore[arg-type]
        return _from_rust_tuple(self._rust.risk_update(patch))  # type: ignore[attr-defined]

    def enter_safe_mode(self) -> RiskConstraints:
        return _from_rust_tuple(self._rust.risk_enter_safe_mode())  # type: ignore[attr-defined]

    def exit_safe_mode(self) -> RiskConstraints:
        return _from_rust_tuple(self._rust.risk_exit_safe_mode())  # type: ignore[attr-defined]

    def halt_trading(self, reason: str = "") -> RiskConstraints:
        del reason  # Reason is logged by callers; cache just records the flag.
        return _from_rust_tuple(self._rust.risk_halt_trading())  # type: ignore[attr-defined]

    def resume_trading(self) -> RiskConstraints:
        return _from_rust_tuple(self._rust.risk_resume_trading())  # type: ignore[attr-defined]


def make_rust_cache() -> Optional[_RustFastRiskCache]:
    """Return a Rust-backed risk cache if the ``dixvision_py_system``
    wheel is importable; otherwise ``None``.

    The underlying Rust cache is a process singleton, so any number
    of wrappers handed out from this factory share the same state.
    That's the intended semantics — governance and Indira are both
    meant to see the same snapshot. Tests that need isolation use
    ``make_python_cache()``."""
    try:
        import dixvision_py_system as _rust  # type: ignore[import-not-found]
    except ImportError:
        return None
    return _RustFastRiskCache(_rust)


# ``FastRiskCache`` is the public name external callers (governance,
# Indira, tests) instantiate. Aliasing it to the Python backend keeps
# the import surface unchanged from the pre-polyglot era; the Rust
# backend is selected through ``get_risk_cache()`` below.
FastRiskCache = _PythonFastRiskCache


# ---------------------------------------------------------------------------
# Module-level default. Picks the Rust backend if the wheel was built
# and installed; otherwise the pure-Python reference. Both are always
# available via the ``make_*_cache`` factories above.
# ---------------------------------------------------------------------------


_rust_impl: Optional[_RustFastRiskCache] = make_rust_cache()
_BACKEND_NAME: str = "rust" if _rust_impl is not None else "python"
_cache: Optional[object] = None
_init_lock = RLock()


def get_risk_cache() -> object:
    """Process-wide risk cache. First call constructs the cache using
    whichever backend was selected at import time; subsequent calls
    return the same instance.

    Return type is intentionally ``object`` rather than
    ``FastRiskCache`` because at runtime the instance may be the
    Rust-backed wrapper; both expose the same duck-typed methods
    (``get``, ``update``, ``enter_safe_mode``, ``exit_safe_mode``,
    ``halt_trading``, ``resume_trading``). Callers interact through
    these methods, not through ``isinstance`` checks."""
    global _cache
    if _cache is None:
        with _init_lock:
            if _cache is None:
                _cache = _rust_impl if _rust_impl is not None else make_python_cache()
    return _cache


def backend() -> str:
    """Which backend is active in this process. Test-hook only —
    runtime code MUST NOT branch on the return value. Returns either
    ``"rust"`` or ``"python"``."""
    return _BACKEND_NAME

File: rust/system/src/fast_risk_cache.rs
//! DIX VISION v42.2 — T0-1 Fast Risk Cache (Rust port).
//!
//! The **only** runtime interface from governance (sole writer) to
//! Indira (hot-path reader). Governance asynchronously publishes
//! precomputed [`RiskConstraints`]; Indira consumes them every tick
//! via a lock-free atomic read.
//!
//! # Polyglot note
//!
//! This is the baseline port — it matches the pre-T0-1 public API
//! byte-for-byte so the `PyO3` seam can expose the same surface as the
//! Python reference impl without breaking any caller. The T0-1
//! staleness-halt / `version_id` / `RiskReading` extensions land in a
//! follow-up PR once the corresponding Python-side PR (#12) is in the
//! base branch — then both sides evolve together.
//!
//! # Concurrency
//!
//! * **Writes** go through a `parking_lot::Mutex`: governance is the
//!   sole writer, so contention is structurally bounded to one thread
//!   bumping the reference + copying a small struct.
//! * **Reads** are lock-free: we store the current `Arc<RiskConstraints>`
//!   inside an `ArcSwap`, and a reader does a single atomic pointer
//!   load + ref-count bump. The SLO p99 < 5 ms budget is dominated by
//!   network, not this module.
//! * The reader contract is explicitly **"eventually consistent up to
//!   the most recent completed `update()`"**: a reader that observes
//!   an older snapshot and a reader that observes the new one are
//!   both valid; governance publishes at a cadence faster than
//!   Indira's tick, so the window is bounded.
//!
//! # Invariants (tested)
//!
//! * `get()` always returns the most recent completed `update()`
//!   (writer holds the lock across the `ArcSwap::store`, so the
//!   publish is atomic from the reader's perspective).
//! * `halt_trading()` monotonically sets `trading_allowed = false`;
//!   only `resume_trading()` can re-enable it.
//! * `enter_safe_mode()` / `exit_safe_mode()` flip `safe_mode` in a
//!   single commit (no intermediate half-state visible to readers).

use arc_swap::ArcSwap;
use parking_lot::Mutex;
use std::sync::Arc;

use crate::time_source;

/// Pre-computed risk limits. Consumed by Indira's fast path.
///
/// Every field mirrors the Python dataclass — changes here must be
/// reflected in the `PyO3` seam (`rust/py_system`) and the Python
/// wrapper (`system/fast_risk_cache.py`) in the same PR.
#[derive(Debug, Clone, PartialEq)]
pub struct RiskConstraints {
    /// Maximum position as fraction of portfolio (1.0 = 100%).
    pub max_position_pct: f64,
    /// Absolute per-order cap, USD.
    pub max_order_size_usd: f64,
    /// Upper volatility band.
    pub volatility_band_high: f64,
    /// Lower volatility band.
    pub volatility_band_low: f64,
    /// Portfolio-level drawdown circuit breaker.
    pub circuit_breaker_drawdown: f64,
    /// Per-trade loss-percentage circuit breaker.
    pub circuit_breaker_loss_pct: f64,
    /// Governance-controlled master switch. `false` halts all trading.
    pub trading_allowed: bool,
    /// Governance-controlled safe-mode flag. `true` rejects trades.
    pub safe_mode: bool,
    /// ISO-8601 UTC timestamp of the last update (informational only —
    /// the monotonic-clock stamp lives on `TimeSource`).
    pub last_updated_utc: String,
}

impl Default for RiskConstraints {
    fn default() -> Self {
        Self {
            max_position_pct: 1.0,
            max_order_size_usd: 10_000.0,
            volatility_band_high: 0.05,
            volatility_band_low: 0.001,
            circuit_breaker_drawdown: 0.04,
            circuit_breaker_loss_pct: 0.01,
            trading_allowed: true,
            safe_mode: false,
            last_updated_utc: String::new(),
        }
    }
}

/// Verdict returned by [`RiskConstraints::allows_trade`].
#[derive(Debug, Clone, PartialEq, Eq)]
pub struct TradeVerdict {
    /// Whether the trade is allowed.
    pub allowed: bool,
    /// Machine-readable reason string. Stable across versions —
    /// decision records stamp this verbatim.
    pub reason: String,
}

impl RiskConstraints {
    /// Check whether a trade of `size_usd` against a portfolio of
    /// `portfolio_usd` is allowed.
    ///
    /// Fail-closed semantics:
    /// * `trading_allowed = false` → reject with `"trading_not_allowed"`.
    /// * `safe_mode = true`        → reject with `"safe_mode_active"`.
    /// * `portfolio_usd <= 0`      → reject with `"portfolio_usd_required"`
    ///   (cannot enforce the percentage circuit breaker without it).
    /// * `size_usd > max_order_size_usd` → reject with a formatted reason.
    /// * `size_usd / portfolio_usd > circuit_breaker_loss_pct` → reject.
    #[must_use]
    pub fn allows_trade(&self, size_usd: f64, portfolio_usd: f64) -> TradeVerdict {
        if !self.trading_allowed {
            return TradeVerdict {
                allowed: false,
                reason: "trading_not_allowed".to_string(),
            };
        }
        if self.safe_mode {
            return TradeVerdict {
                allowed: false,
                reason: "safe_mode_active".to_string(),
            };
        }
        if portfolio_usd <= 0.0 {
            return TradeVerdict {
                allowed: false,
                reason: "portfolio_usd_required".to_string(),
            };
        }
        if size_usd > self.max_order_size_usd {
            return TradeVerdict {
                allowed: false,
                reason: format!(
                    "size_usd_{:.2}_exceeds_max_{:.2}",
                    size_usd, self.max_order_size_usd
                ),
            };
        }
        let pct = size_usd / portfolio_usd;
        if pct > self.circuit_breaker_loss_pct {
            return TradeVerdict {
                allowed: false,
                reason: format!(
                    "size_pct_{:.4}_exceeds_limit_{}",
                    pct, self.circuit_breaker_loss_pct
                ),
            };
        }
        TradeVerdict {
            allowed: true,
            reason: "ok".to_string(),
        }
    }
}

/// Patch record applied by `FastRiskCache::update`. Every field is
/// optional: `None` means "keep the previous value", `Some(x)` means
/// "replace with x".
#[derive(Debug, Default, Clone)]
pub struct RiskUpdate {
    /// Override `max_position_pct`.
    pub max_position_pct: Option<f64>,
    /// Override `max_order_size_usd`.
    pub max_order_size_usd: Option<f64>,
    /// Override `volatility_band_high`.
    pub volatility_band_high: Option<f64>,
    /// Override `volatility_band_low`.
    pub volatility_band_low: Option<f64>,
    /// Override `circuit_breaker_drawdown`.
    pub circuit_breaker_drawdown: Option<f64>,
    /// Override `circuit_breaker_loss_pct`.
    pub circuit_breaker_loss_pct: Option<f64>,
    /// Override `trading_allowed`.
    pub trading_allowed: Option<bool>,
    /// Override `safe_mode`.
    pub safe_mode: Option<bool>,
}

/// Atomic single-writer, multi-reader risk cache.
///
/// Governance is the sole writer (async). Indira reads every tick
/// with zero lock contention.
pub struct FastRiskCache {
    /// Canonical pointer. Readers `load()`; writers `store()` inside
    /// the mutex to serialize ordering.
    current: ArcSwap<RiskConstraints>,
    /// Serializes governance writes. Reads do NOT take this lock.
    write_lock: Mutex<()>,
}

impl FastRiskCache {
    /// Construct a cache with default constraints.
    #[must_use]
    pub fn new() -> Self {
        let initial = RiskConstraints {
            last_updated_utc: current_utc_iso(),
            ..Default::default()
        };
        Self {
            current: ArcSwap::from(Arc::new(initial)),
            write_lock: Mutex::new(()),
        }
    }

    /// Lock-free read of the current constraints snapshot. The caller
    /// receives a cheap `Arc<RiskConstraints>` — no copy.
    #[must_use]
    pub fn get(&self) -> Arc<RiskConstraints> {
        self.current.load_full()
    }

    /// Apply a `RiskUpdate`, stamping `last_updated_utc` atomically.
    /// Returns the new snapshot.
    pub fn update(&self, patch: &RiskUpdate) -> Arc<RiskConstraints> {
        let _guard = self.write_lock.lock();
        let prev = self.current.load_full();
        let mut next = (*prev).clone();
        if let Some(v) = patch.max_position_pct {
            next.max_position_pct = v;
        }
        if let Some(v) = patch.max_order_size_usd {
            next.max_order_size_usd = v;
        }
        if let Some(v) = patch.volatility_band_high {
            next.volatility_band_high = v;
        }
        if let Some(v) = patch.volatility_band_low {
            next.volatility_band_low = v;
        }
        if let Some(v) = patch.circuit_breaker_drawdown {
            next.circuit_breaker_drawdown = v;
        }
        if let Some(v) = patch.circuit_breaker_loss_pct {
            next.circuit_breaker_loss_pct = v;
        }
        if let Some(v) = patch.trading_allowed {
            next.trading_allowed = v;
        }
        if let Some(v) = patch.safe_mode {
            next.safe_mode = v;
        }
        next.last_updated_utc = current_utc_iso();
        let arc = Arc::new(next);
        self.current.store(Arc::clone(&arc));
        arc
    }

    /// Convenience: halt trading (`trading_allowed = false`, `safe_mode = true`).
    pub fn enter_safe_mode(&self) -> Arc<RiskConstraints> {
        self.update(&RiskUpdate {
            safe_mode: Some(true),
            trading_allowed: Some(false),
            ..Default::default()
        })
    }

    /// Exit safe-mode: re-enable trading. Callers must still verify
    /// governance-level policy before this is executed.
    pub fn exit_safe_mode(&self) -> Arc<RiskConstraints> {
        self.update(&RiskUpdate {
            safe_mode: Some(false),
            trading_allowed: Some(true),
            ..Default::default()
        })
    }

    /// Halt trading without entering safe-mode (for operator stop-all
    /// that can be reversed without a full safe-mode exit dance).
    pub fn halt_trading(&self) -> Arc<RiskConstraints> {
        self.update(&RiskUpdate {
            trading_allowed: Some(false),
            ..Default::default()
        })
    }

    /// Resume trading. Clears `safe_mode` as well so operators don't
    /// have to remember to un-set both independently.
    pub fn resume_trading(&self) -> Arc<RiskConstraints> {
        self.update(&RiskUpdate {
            trading_allowed: Some(true),
            safe_mode: Some(false),
            ..Default::default()
        })
    }
}

impl Default for FastRiskCache {
    fn default() -> Self {
        Self::new()
    }
}

/// ISO-8601 UTC timestamp derived from the canonical `TimeSource`.
///
/// Format: `YYYY-MM-DDTHH:MM:SS.ffffff+00:00` — matches the shape
/// produced by Python's `datetime.datetime.isoformat()` on a timezone-
/// aware UTC datetime, so governance / Indira consumers that compare
/// strings across the seam see the same bytes.
///
/// We use `chrono` only for the civil-calendar conversion; the actual
/// wall-clock reading comes from `TimeSource` (our canonical clock).
fn current_utc_iso() -> String {
    let stamp = time_source::now();
    let utc_ns = stamp.utc_nanos;
    let secs = i64::try_from(utc_ns.div_euclid(1_000_000_000)).unwrap_or(0);
    let sub_nanos = u32::try_from(utc_ns.rem_euclid(1_000_000_000)).unwrap_or(0);
    let dt = chrono::DateTime::<chrono::Utc>::from_timestamp(secs, sub_nanos).unwrap_or_default();
    // "%.6f" emits the `.ffffff` fractional-seconds piece; "%:z"
    // emits "+00:00" for UTC. Together they match Python's
    // `datetime.isoformat()` for aware UTC datetimes.
    dt.format("%Y-%m-%dT%H:%M:%S%.6f%:z").to_string()
}

#[cfg(test)]
#[allow(
    clippy::expect_used,
    clippy::unwrap_used,
    clippy::uninlined_format_args
)]
mod tests {
    use super::*;
    use std::sync::Arc;
    use std::thread;

    #[test]
    fn defaults_are_permissive_and_safe() {
        let c = RiskConstraints::default();
        assert!(c.trading_allowed);
        assert!(!c.safe_mode);
        assert!((c.max_position_pct - 1.0).abs() < f64::EPSILON);
        assert!((c.max_order_size_usd - 10_000.0).abs() < f64::EPSILON);
        assert_eq!(c.last_updated_utc, "");
    }

    #[test]
    fn allows_trade_rejects_when_trading_disabled() {
        let c = RiskConstraints {
            trading_allowed: false,
            ..Default::default()
        };
        let v = c.allows_trade(100.0, 10_000.0);
        assert!(!v.allowed);
        assert_eq!(v.reason, "trading_not_allowed");
    }

    #[test]
    fn allows_trade_rejects_when_safe_mode() {
        let c = RiskConstraints {
            safe_mode: true,
            ..Default::default()
        };
        let v = c.allows_trade(100.0, 10_000.0);
        assert!(!v.allowed);
        assert_eq!(v.reason, "safe_mode_active");
    }

    #[test]
    fn allows_trade_requires_portfolio() {
        let c = RiskConstraints::default();
        let v = c.allows_trade(100.0, 0.0);
        assert!(!v.allowed);
        assert_eq!(v.reason, "portfolio_usd_required");
        let v2 = c.allows_trade(100.0, -5.0);
        assert!(!v2.allowed);
        assert_eq!(v2.reason, "portfolio_usd_required");
    }

    #[test]
    fn allows_trade_enforces_absolute_cap_before_percentage() {
        let c = RiskConstraints::default();
        // 20_000 > max_order_size_usd (10_000) — must reject on the
        // absolute cap regardless of how large the portfolio is.
        let v = c.allows_trade(20_000.0, 1_000_000_000.0);
        assert!(!v.allowed);
        assert!(v.reason.starts_with("size_usd_"), "got {}", v.reason);
        assert!(v.reason.contains("exceeds_max_"), "got {}", v.reason);
    }

    #[test]
    fn allows_trade_enforces_percentage_circuit_breaker() {
        let c = RiskConstraints::default();
        // 2% of 10_000 portfolio = 200; circuit_breaker_loss_pct = 1%.
        let v = c.allows_trade(200.0, 10_000.0);
        assert!(!v.allowed);
        assert!(v.reason.starts_with("size_pct_"), "got {}", v.reason);
    }

    #[test]
    fn allows_trade_accepts_within_limits() {
        let c = RiskConstraints::default();
        // 0.5% of a 10_000 portfolio = 50; both caps pass.
        let v = c.allows_trade(50.0, 10_000.0);
        assert!(v.allowed);
        assert_eq!(v.reason, "ok");
    }

    #[test]
    fn cache_get_returns_defaults_on_fresh() {
        let cache = FastRiskCache::new();
        let snap = cache.get();
        assert!(snap.trading_allowed);
        assert!(!snap.safe_mode);
        assert!(!snap.last_updated_utc.is_empty());
    }

    #[test]
    fn cache_update_swaps_reference_atomically() {
        let cache = FastRiskCache::new();
        let before = cache.get();
        let after = cache.update(&RiskUpdate {
            max_order_size_usd: Some(42_000.0),
            ..Default::default()
        });
        let observed = cache.get();
        assert!(!Arc::ptr_eq(&before, &after));
        assert!(Arc::ptr_eq(&after, &observed));
        assert!((after.max_order_size_usd - 42_000.0).abs() < f64::EPSILON);
    }

    #[test]
    fn cache_enter_safe_mode_halts_trading() {
        let cache = FastRiskCache::new();
        let s = cache.enter_safe_mode();
        assert!(s.safe_mode);
        assert!(!s.trading_allowed);
        let v = s.allows_trade(10.0, 10_000.0);
        assert!(!v.allowed);
    }

    #[test]
    fn cache_exit_safe_mode_resumes_trading() {
        let cache = FastRiskCache::new();
        cache.enter_safe_mode();
        let s = cache.exit_safe_mode();
        assert!(!s.safe_mode);
        assert!(s.trading_allowed);
    }

    #[test]
    fn cache_halt_and_resume_cycle() {
        let cache = FastRiskCache::new();
        let h = cache.halt_trading();
        assert!(!h.trading_allowed);
        let r = cache.resume_trading();
        assert!(r.trading_allowed);
        assert!(!r.safe_mode);
    }

    #[test]
    fn concurrent_readers_never_see_partial_update() {
        // Writer flips `max_order_size_usd` between two wildly
        // different values. Readers must always observe one of the
        // two — never an intermediate or torn value.
        const READER_COUNT: usize = 8;
        const ITERATIONS: usize = 2_000;
        let cache = Arc::new(FastRiskCache::new());

        let writer_cache = Arc::clone(&cache);
        let writer = thread::spawn(move || {
            for i in 0..ITERATIONS {
                let v = if i & 1 == 0 { 10_000.0 } else { 99_999.0 };
                writer_cache.update(&RiskUpdate {
                    max_order_size_usd: Some(v),
                    ..Default::default()
                });
            }
        });

        let readers: Vec<_> = (0..READER_COUNT)
            .map(|_| {
                let c = Arc::clone(&cache);
                thread::spawn(move || {
                    for _ in 0..ITERATIONS {
                        let snap = c.get();
                        let v = snap.max_order_size_usd;
                        assert!(
                            (v - 10_000.0).abs() < f64::EPSILON
                                || (v - 99_999.0).abs() < f64::EPSILON,
                            "torn read: {}",
                            v
                        );
                    }
                })
            })
            .collect();

        writer.join().expect("writer thread panicked");
        for r in readers {
            r.join().expect("reader thread panicked");
        }
    }

    #[test]
    fn last_updated_utc_is_iso_8601() {
        let cache = FastRiskCache::new();
        let s = cache.get();
        // Shape: YYYY-MM-DDTHH:MM:SS.ffffff+00:00  (32 chars)
        assert_eq!(
            s.last_updated_utc.len(),
            32,
            "unexpected ISO length: {}",
            s.last_updated_utc
        );
        assert_eq!(&s.last_updated_utc[4..5], "-");
        assert_eq!(&s.last_updated_utc[7..8], "-");
        assert_eq!(&s.last_updated_utc[10..11], "T");
        assert_eq!(&s.last_updated_utc[13..14], ":");
        assert_eq!(&s.last_updated_utc[19..20], ".");
        assert!(s.last_updated_utc.ends_with("+00:00"));
    }

    #[test]
    fn utc_iso_matches_chrono_shape_on_epoch() {
        let dt = chrono::DateTime::<chrono::Utc>::from_timestamp(0, 0).unwrap();
        let expected = dt.format("%Y-%m-%dT%H:%M:%S%.6f%:z").to_string();
        assert_eq!(expected, "1970-01-01T00:00:00.000000+00:00");
    }
}
